Implement command generator for arbitrary pipelines

Aratz opened the pull request:
**What problems does this PR solve?**
This PR is the first step towards enabling the service to run multiple pipelines. It updates the nextflow command generator so that it can run arbitrary pipelines from a pipeline config file, which format is specified in the following schema: https://github.com/arteria-project/sequencing-report-service/blob/53b153f195b784aa2a2cc3be0a1dd147c7ba54ac/config/pipeline_config/schema.json.

Nf-core pipelines use a specific input samplesheet to determine what to process (see [nf-core/demultiplex](https://nf-co.re/demultiplex/1.4.0/docs/usage#samplesheet-input) for instance). The service handles this by writing down such samplesheet in the runfolder to be processed. The content of this samplesheet can either be provided in the pipeline's configuration file or as a parameter to the handler.

**How has the changes been tested?**
Tests have been updated, full validation should check both old and new pipelines.

**Reasons for careful code review**
If any of the boxes below are checked, extra careful code review should be initiated.

  - [ ] This PR contains code that could remove data

Aratz committed:
Implement command generator for arbitrary pipelines

matrulda commented on config/pipeline_config/schema.json:
```diff
@@ -0,0 +1,24 @@
+      "type": "object",
+      "additionalProperties": {"type": "string"}
+    },
+    "parameters": {
```
If we call this `pipeline_parameters`, we could also have `nextflow_parameters` to easily pass extra nextflow parameters like `-with-tower` etc.

Aratz replied:
Good point! Will fix, thanks for the feedback!

matrulda reviewed (commented):
Sweet, looks great! The Nextflow module got so much cleaner. Left a suggestion you can have a look at.

Aratz committed:
Make `parameters` and `environment` optional

Aratz committed:
Put nextflow parameters in their own section

Aratz committed:
Add support for nf-core input samplesheets

Most nf-core pipelines define input data in a pipeline-specific
samplesheet. This commit makes it possible to provide such a samplesheet
to the pipeline.

The content of the samplesheet can be either provided directly to the
endpoint or can be described in the config file (in this case it will
usually make use of {runfolder_name} and {runfolder_path}.

Aratz committed:
Connect handlers

Aratz committed:
Update ci

b97pla commented on sequencing_report_service/nextflow.py:
```diff
@@ -2,106 +2,174 @@
+        cmd += [
+            arg
+            for key, value in config.get("pipeline_parameters", {}).items()
+            for arg in [f"--{key}", f"{value}"]
```
pipeline parameters can also be flags, so I'd expect that you need to have a similar construct as for nextflow parameters

b97pla commented on sequencing_report_service/nextflow.py:
```diff
@@ -2,106 +2,174 @@
+            "runfolder_name": runfolder_path.name,
+        }
+
+        input_samplesheet = input_samplesheet or config.get("input_samplesheet", "")
```
isn't it a bit clunky to pass the samplesheet content around as a string? I guess I'm also a bit unsure about what this will contain. for the analysis pipelines, it is quite extensive, typically with one entry per sample and it can be quite big.

Aratz replied:
Yeah I agree, although I don't really have a better solution. Are the samplesheets for the analysis pipelines much bigger than the runfolder samplesheets? I think there is the same issue in arteria-bcl2fastq and the same solution has been implemented there (see https://github.com/arteria-project/arteria-bcl2fastq/blob/afb1332c016d7af99cb710d3c6f4fe8f10775422/bcl2fastq/handlers/bcl2fastq_handlers.py#L165)

b97pla replied:
yeah, I couldn't really suggest a better approach either, unless implementing a lot of extra functionality to handle this. In theory (and sometimes in practice), the analysis sample sheets can. be much bigger since they can combine samples from multiple runfolders. But I think that's only a problem in theory, in reality I don't think it will be a problem.

b97pla commented on sequencing_report_service/nextflow.py:
```diff
@@ -2,106 +2,174 @@
+            "current_year": datetime.datetime.now().year,
+            "runfolder_path": str(runfolder_path),
+            "runfolder_name": runfolder_path.name,
+        }
```
I guess a future improvement could be to be able to also pass overriding defaults to the service. I assume that especially `current_year` could be something that needs overriding.

Aratz replied:
Good point, I guess in this case the user would provide a new `defaults` dictionary to the service. I'll keep that in mind for later :)

b97pla commented on sequencing_report_service/nextflow.py:
```diff
@@ -2,106 +2,174 @@
+                f.write(input_samplesheet)
+            defaults["input_samplesheet"] = input_samplesheet_path
+
+        return interpolate_variables(config, defaults)
```
I think it may be a bit confusing to having `input_samplesheet` used in `config` both as a key with default content (as a string) and as a keyword `{input_samplesheet}` that will be substituted for a path. perhaps they can be named differently to make the difference clear?

Aratz replied:
Good point! I'll think about something else

b97pla commented on sequencing_report_service/nextflow.py:
```diff
@@ -2,106 +2,174 @@
+        -------
+        config: dict
+       """
         try:
```
personally, I would also move the schema validation to a separate function 😃

Aratz replied:
I thought a little bit more about it now and I think keeping the config and the validation together prevents unvalidated configs to be loaded up in the future (e.g. if one misses that there is a validation function). So I think I'll keep them together for now.

I did however reorganized the whole `nextflow` module, e.g. I don't think we need a class in this module for instance, hope it makes the whole file clearer

b97pla commented on sequencing_report_service/services/local_runner_service.py:
```diff
@@ -102,14 +102,30 @@ async def _start_process(self, job_id):
+            )
+
+            if ext_args:
+                nf_cmd['command'] += ext_args
```
would it be more logical to handle the `ext_args` in the `self._nextflow_jobs_factory.command` function instead?

b97pla reviewed (approved):
:+1: great work, I think the code looks good

I had a couple of questions/comments that you can address if you want. Or not.

Aratz committed:
Add support for flag parameters

Aratz committed:
Refactor nextflow module

Aratz commented:
Great review! I fixed most things but I don't really have a solution for the samplesheet content being passed around as a string, let me know if you have any suggestion, otherwise I think it should be good to go

b97pla commented:
:+1: great work, I'm happy
